apps/els_dap/src/els_dap_rpc.erl

```diff
@@ -74,25 +76,71 @@ eval(Node, Input, Bindings) ->
-                end;
-            IntSource ->
-                {ok, IntSource}
-        end,
-    case MaybeSource of
-        {ok, Source} ->
-            Source;
+    case file_from_module_info(Node, Module) of
+        {ok, FileFromInt} ->
+            {ok, FileFromInt};
         {error, not_found} ->
-            CompileOpts = module_info(Node, Module, compile),
-            proplists:get_value(source, CompileOpts)
+            case file_from_int(Node, Module) of
+                {ok, FileFromModuleInfo} ->
+                    {ok, FileFromModuleInfo};
+                {error, not_found} ->
+                    file_from_code_server(Node, Module)
+            end
+    end.
+
+-spec file_from_int(node(), module()) -> {ok, file:filename()} | {error, not_found}.
+file_from_int(Node, Module) ->
+    ?LOG_DEBUG("Looking in Int: [~p]", [Module]),
+    case rpc:call(Node, int, file, [Module]) of
+        {error, not_loaded} ->
+            ?LOG_DEBUG("Not Found in Int: [~p]", [Module]),
+            {error, not_found};
+        Path ->
+            ?LOG_DEBUG("Found in Int: [~p]", [Path]),
+            {ok, Path}
+    end.
+
+-spec file_from_code_server(node(), module()) -> {ok, file:filename()} | {error, not_found}.
+file_from_code_server(Node, Module) ->
+    ?LOG_DEBUG("Looking in Code Server: [~p]", [Module]),
+    BeamName = atom_to_list(Module) ++ ".beam",
+    case rpc:call(Node, code, where_is_file, [BeamName]) of
+        non_existing ->
+            ?LOG_DEBUG("Not found in Code Server: [~p]", [Module]),
+            {error, not_found};
+        BeamFile ->
+            ?LOG_DEBUG("Found in Code Server: [~p]", [BeamFile]),
+            rpc:call(Node, filelib, find_source, [BeamFile])
+    end.
+
+-spec file_from_module_info(node(), module()) -> {ok, file:filename()} | {error, not_found}.
+file_from_module_info(Node, Module) ->
+    ?LOG_DEBUG("Looking in Module Info: [~p]", [Module]),
+    CompileOpts = module_info(Node, Module, compile),
+    case proplists:get_value(source, CompileOpts) of
+        undefined ->
+            ?LOG_DEBUG("Not found in Module Info: [~p]", [Module]),
+            {error, not_found};
+        Source ->
+            ?LOG_DEBUG("Found in Module Info: [~p]", [Source]),
+            case rpc:call(Node, filelib, is_regular, [Source]) of
+                true ->
+                    {ok, Cwd} = rpc:call(Node, file, get_cwd, []),
+                    case rpc:call(Node, filelib, safe_relative_path, [Source, Cwd]) of
+                        unsafe ->
+                            %% File is already absolute
+                            {ok, Source};
+                        RelativePath ->
+                            Joined = filename:join(Cwd, RelativePath),
+                            ?LOG_DEBUG("Composed Absolute Path as: [~p]", [Joined]),
+                            {ok, Joined}
+                    end;
+                false ->
+                    %% Build is not hermetic
```

Comment:
it might be hermetic, e.g. with +deterministic you get only the base name and is_regular will fail.

Reply:
When I tried the `+determinsitc` option on OTP 25 I didn't get a `source` entry at all, so that case would be handled in the `undefined` clause above. But I can remove the comment.